File: GreenQuiz/Hubs/QuizHub.cs
using GreenQuiz.Models;
using Microsoft.AspNetCore.SignalR;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace GreenQuiz.Hubs
{
    public class QuizHub : Hub<IQuizHub>
    {
        public override Task OnConnected()
        {
            return Clients.Client(Context.ConnectionId).SetConnectionId(Context.ConnectionId);
        }
        // Server side methods called from client
        public Task Subscribe(string personId)
        {
            return Groups.Add(Context.ConnectionId, personId);
        }

        public Task Unsubscribe(string personId)
        {
            return Groups.Remove(Context.ConnectionId, personId);
        }

        public Task SubscribeToSession(string sessionId)
        {
            return Groups.Add(Context.ConnectionId, sessionId);
        }
        public Task UnsubscribeFromSession(string sessionId)
        {
            return Groups.Remove(Context.ConnectionId, sessionId);
        }
    }
}

public interface IQuizHub
{
    Task AddQuiz(Quiz quiz);
    Task UpdateSession(QuizSession quizSession);
    Task SetConnectionId(string connectionId);
}

File: GreenQuiz/ClientApp/app/model/quizhub.ts
/// <reference path="../../../typings/globals/signalr/index.d.ts" />
import { Quiz } from './quiz.model';
import { QuizSession } from './quizsession.model';
//interface SignalR {
//    quizHub: HubProxy;
//}
// export interface HubProxy {
//    client: IQuizHubClient;
//}
//interface IQuizHubClient {
//    publishQuiz(quiz: Quiz);
// }
//interface HubConnection extends SignalR {
//    hub: SignalR.Hub.Connection;
//}


export interface QuizSignalR extends SignalR {
    quizHub: QuizProxy
}

export interface QuizProxy {
    client: QuizClient;
    server: FeedServer;
}

export interface QuizClient {
    addQuiz(quiz: Quiz): void;
    updateSession(session: QuizSession): void;
}

export interface FeedServer {
    subscribe(personId: string): void;
    unsubscribe(personId: string): void;
    subscribeToSession(sessionId: string): void;
    unsubscribeFromSession(sessionId: string): void;
}

export enum SignalRConnectionStatus {
    Connected = 1,
    Disconnected = 2,
    Error = 3
}
